lib/teleterm/services/userpreferences/userpreferences.go

```diff
@@ -23,43 +23,95 @@ import (
 	}
+	rootPreferences := rootPreferencesResponse.GetPreferences()
+
+	var leafPreferences *userpreferencesv1.UserPreferences
+	if leafClient != nil {
+		preferences, err := leafClient.GetUserPreferences(ctx, &userpreferencesv1.GetUserPreferencesRequest{})
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		leafPreferences = preferences.GetPreferences()
+	}
+
+	clusterPreferences := rootPreferences.GetClusterPreferences()
+	if leafPreferences != nil {
+		clusterPreferences = leafPreferences.GetClusterPreferences()
+	}
+
 	return &api.UserPreferences{
-		ClusterPreferences:         preferences.GetPreferences().ClusterPreferences,
-		UnifiedResourcePreferences: preferences.GetPreferences().UnifiedResourcePreferences,
+		UnifiedResourcePreferences: rootPreferences.UnifiedResourcePreferences,
+		ClusterPreferences:         clusterPreferences,
 	}, nil
 }
 
 // Update updates the preferences for a given user.
 // Only the properties that are set (cluster_preferences, unified_resource_preferences) will be updated.
-func Update(ctx context.Context, client Client, newPreferences *api.UserPreferences) error {
+func Update(ctx context.Context, rootClient Client, leafClient Client, newPreferences *api.UserPreferences) (*api.UserPreferences, error) {
 	// We have to fetch the full user preferences struct and modify only
 	// the fields that change.
 	// Calling `UpsertUserPreferences` with only the modified values would reset
 	// the rest of the preferences.
-	currentPreferencesResponse, err := client.GetUserPreferences(ctx, &userpreferencesv1.GetUserPreferencesRequest{})
+	rootPreferencesResponse, err := rootClient.GetUserPreferences(ctx, &userpreferencesv1.GetUserPreferencesRequest{})
 	if err != nil {
-		return trace.Wrap(err)
+		return nil, trace.Wrap(err)
 	}
+	rootPreferences := rootPreferencesResponse.GetPreferences()
 
-	currentPreferences := currentPreferencesResponse.GetPreferences()
+	var leafPreferences *userpreferencesv1.UserPreferences
+	if leafClient != nil {
+		response, err := leafClient.GetUserPreferences(ctx, &userpreferencesv1.GetUserPreferencesRequest{})
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		leafPreferences = response.GetPreferences()
+	}
 
-	if newPreferences.UnifiedResourcePreferences != nil {
-		currentPreferences.UnifiedResourcePreferences = newPreferences.UnifiedResourcePreferences
+	hasUnifiedResourcePreferencesForRoot := newPreferences.UnifiedResourcePreferences != nil
+	hasClusterPreferencesForRoot := newPreferences.ClusterPreferences != nil && leafPreferences == nil
+
+	if hasUnifiedResourcePreferencesForRoot || hasClusterPreferencesForRoot {
+		if hasUnifiedResourcePreferencesForRoot {
+			rootPreferences.UnifiedResourcePreferences = newPreferences.UnifiedResourcePreferences
+		}
+		if hasClusterPreferencesForRoot {
+			rootPreferences.ClusterPreferences = newPreferences.ClusterPreferences
+		}
+
+		err = rootClient.UpsertUserPreferences(ctx, &userpreferencesv1.UpsertUserPreferencesRequest{
```

Comment:
This code is within a new scope, so we can do `err :=` instead, just to be safe.

It's not particularly risky here, but writing code like this bit me a couple of times when writing parallel tests (see https://github.com/gravitational/teleport/pull/29803).

Reply:
Not super related, but just reminded me of this other `err` thing I ran into once by returning an "old" error on accident. https://github.com/gravitational/teleport/pull/31003

Like I said, not related to this but just a funny story.

Reply:
Thanks, done :)